File: community/cypher/src/test/scala/org/neo4j/cypher/ExecutionResultTest.scala
/**
 * Copyright (c) 2002-2013 "Neo Technology,"
 * Network Engine for Objects in Lund AB [http://neotechnology.com]
 *
 * This file is part of Neo4j.
 *
 * Neo4j is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
package org.neo4j.cypher

import org.scalatest.Assertions
import org.junit.Test
import org.junit.Assert._
import java.util.regex.Pattern

class ExecutionResultTest extends ExecutionEngineHelper with Assertions {
  @Test def columnOrderIsPreserved() {
    val columns = List("zero", "one", "two", "three", "four", "five", "six", "seven", "eight", "nine")

    columns.foreach(createNode)

    val q="start zero=node(0), one=node(1), two=node(2), three=node(3), four=node(4), five=node(5), six=node(6), seven=node(7), eight=node(8), nine=node(9)" +
      "return zero, one, two, three, four, five, six, seven, eight, nine"

    val result = execute(q)

    assert( result.columns === columns )
    val regex = "zero.*one.*two.*three.*four.*five.*six.*seven.*eight.*nine"
    val pattern = Pattern.compile(regex)

    assertTrue( "Columns did not apperar in the expected order: \n" + result.dumpToString(), pattern.matcher(result.dumpToString()).find() )
  }

  @Test def correctLabelStatisticsForCreate() {
    val result = execute("create (n:foo:bar)")
    val stats  = result.queryStatistics()

    assert(stats.labelsAdded === 2)
    assert(stats.labelsRemoved === 0)
  }

  @Test def correctLabelStatisticsForAdd() {
    val n      = createNode()
    val result = execute(s"start n=node(${n.getId}) set n:foo:bar")
    val stats  = result.queryStatistics()

    assert(stats.labelsAdded === 2)
    assert(stats.labelsRemoved === 0)
  }

  @Test def correctLabelStatisticsForRemove() {
    val n      = createNode()
    execute(s"start n=node(${n.getId}) set n:foo:bar")
    val result = execute(s"start n=node(${n.getId}) remove n:foo:bar")
    val stats  = result.queryStatistics()

    assert(stats.labelsAdded === 0)
    assert(stats.labelsRemoved === 2)
  }

  @Test def correctLabelStatisticsForAddAndRemove() {
    val n      = createLabeledNode("foo", "bar")
    val result = execute(s"start n=node(${n.getId}) set n:baz remove n:foo:bar")
    val stats  = result.queryStatistics()

    assert(stats.labelsAdded === 1)
    assert(stats.labelsRemoved === 2)
  }


  @Test def correctLabelStatisticsForLabelAddedTwice() {
    val n      = createLabeledNode("foo", "bar")
    val result = execute(s"start n=node(${n.getId}) set n:bar:baz")
    val stats  = result.queryStatistics()

    assert(stats.labelsAdded === 1)
    assert(stats.labelsRemoved === 0)
  }

  @Test def correctLabelStatisticsForRemovalOfUnsetLabel() {
    val n      = createLabeledNode("foo", "bar")
    val result = execute(s"start n=node(${n.getId}) remove n:baz:foo")
    val stats  = result.queryStatistics()

    assert(stats.labelsAdded === 0)
    assert(stats.labelsRemoved === 1)
  }

  @Test def correctIndexStatisticsForIndexAdded() {
    val result = execute("create index on :Person(name)")
    val stats  = result.queryStatistics()

    assert(stats.indexesAdded === 1)
    assert(stats.indexesRemoved === 0)
  }

  @Test def correctIndexStatisticsForIndexAddedTwice() {
    execute("create index on :Person(name)")

    val result = execute("create index on :Person(name)")
    val stats  = result.queryStatistics()

    assert(stats.indexesAdded === 0)
    assert(stats.indexesRemoved === 0)
  }

  @Test def correctConstraintStatisticsForUniquenessConstraintAdded() {
    val result = execute("create constraint on (n:Person) assert n.name is unique")
    val stats  = result.queryStatistics()

    assert(stats.constraintsAdded === 1)
    assert(stats.constraintsRemoved === 0)
  }

  @Test def correctConstraintStatisticsForUniquenessConstraintAddedTwice() {
    execute("create constraint on (n:Person) assert n.name is unique")

    val result = execute("create constraint on (n:Person) assert n.name is unique")
    val stats  = result.queryStatistics()

    assert(stats.constraintsAdded === 0)
    assert(stats.constraintsRemoved === 0)
  }
}

File: community/cypher/cypher/src/test/java/org/neo4j/cypher/internal/javacompat/ExecutionResultTest.java
/*
 * Copyright (c) 2002-2018 "Neo Technology,"
 * Network Engine for Objects in Lund AB [http://neotechnology.com]
 *
 * This file is part of Neo4j.
 *
 * Neo4j is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
package org.neo4j.cypher.internal.javacompat;

import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;

import javax.annotation.Resource;

import org.neo4j.cypher.ArithmeticException;
import org.neo4j.graphdb.Node;
import org.neo4j.graphdb.QueryExecutionException;
import org.neo4j.graphdb.ResourceIterator;
import org.neo4j.graphdb.Result;
import org.neo4j.graphdb.Transaction;
import org.neo4j.graphdb.spatial.Point;
import org.neo4j.kernel.api.KernelTransaction;
import org.neo4j.kernel.impl.core.ThreadToStatementContextBridge;
import org.neo4j.kernel.impl.coreapi.TopLevelTransaction;
import org.neo4j.kernel.impl.query.QueryExecutionKernelException;
import org.neo4j.test.extension.ImpermanentDatabaseExtension;
import org.neo4j.test.rule.ImpermanentDatabaseRule;

import static java.lang.Math.round;
import static java.util.Arrays.asList;
import static org.hamcrest.CoreMatchers.equalTo;
import static org.hamcrest.MatcherAssert.assertThat;
import static org.hamcrest.Matchers.instanceOf;
import static org.hamcrest.Matchers.is;
import static org.hamcrest.core.IsNull.notNullValue;
import static org.hamcrest.core.IsNull.nullValue;
import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.neo4j.helpers.collection.MapUtil.map;

@ExtendWith( ImpermanentDatabaseExtension.class )
public class ExecutionResultTest
{
    @Resource
    public ImpermanentDatabaseRule db;

    //TODO this test is not valid for compiled runtime as the transaction will be closed when the iterator was created
    @Test
    public void shouldCloseTransactionsWhenIteratingResults()
    {
        // Given an execution result that has been started but not exhausted
        createNode();
        createNode();
        Result executionResult = db.execute( "CYPHER runtime=interpreted MATCH (n) RETURN n" );
        executionResult.next();
        assertThat( activeTransaction(), is( notNullValue() ) );

        // When
        executionResult.close();

        // Then
        assertThat( activeTransaction(), is( nullValue() ) );
    }

    //TODO this test is not valid for compiled runtime as the transaction will be closed when the iterator was created
    @Test
    public void shouldCloseTransactionsWhenIteratingOverSingleColumn()
    {
        // Given an execution result that has been started but not exhausted
        createNode();
        createNode();
        Result executionResult = db.execute( "CYPHER runtime=interpreted MATCH (n) RETURN n" );
        ResourceIterator<Node> resultIterator = executionResult.columnAs( "n" );
        resultIterator.next();
        assertThat( activeTransaction(), is( notNullValue() ) );

        // When
        resultIterator.close();

        // Then
        assertThat( activeTransaction(), is( nullValue() ) );
    }

    @Test
    public void shouldThrowAppropriateException()
    {
        try
        {
            db.execute( "RETURN rand()/0" ).next();
        }
        catch ( QueryExecutionException ex )
        {
            assertThat( ex.getCause(), instanceOf( QueryExecutionKernelException.class ) );
            assertThat( ex.getCause().getCause(), instanceOf( ArithmeticException.class ) );
        }
    }

    @Test
    public void shouldThrowAppropriateExceptionAlsoWhenVisiting()
    {
        assertThrows( org.neo4j.cypher.ArithmeticException.class, () -> db.execute( "RETURN rand()/0" ).accept( row -> true ) );
    }

    private void createNode()
    {
        try ( Transaction tx = db.beginTx() )
        {
            db.createNode();
            tx.success();
        }
    }

    @Test
    public void shouldHandleListsOfPointsAsInput()
    {
        // Given
        Point point1 =
                (Point) db.execute( "RETURN point({latitude: 12.78, longitude: 56.7}) as point" ).next().get( "point" );
        Point point2 =
                (Point) db.execute( "RETURN point({latitude: 12.18, longitude: 56.2}) as point" ).next().get( "point" );

        // When
        double distance = (double) db.execute( "RETURN distance({points}[0], {points}[1]) as dist",
                map( "points", asList( point1, point2 ) ) ).next().get( "dist" );
        // Then
        assertThat( round( distance ), equalTo( 86107L ) );
    }

    @Test
    public void shouldHandleMapWithPointsAsInput()
    {
        // Given
        Point point1 = (Point) db.execute( "RETURN point({latitude: 12.78, longitude: 56.7}) as point" ).next().get( "point" );
        Point point2 = (Point) db.execute( "RETURN point({latitude: 12.18, longitude: 56.2}) as point" ).next().get( "point" );

        // When
        double distance = (double) db.execute( "RETURN distance({points}['p1'], {points}['p2']) as dist",
                map( "points", map( "p1", point1, "p2", point2 ) ) ).next().get( "dist" );
        // Then
        assertThat( round( distance ), equalTo( 86107L ) );
    }

    private TopLevelTransaction activeTransaction()
    {
        ThreadToStatementContextBridge bridge =
                db.getDependencyResolver().resolveDependency( ThreadToStatementContextBridge.class );
        KernelTransaction kernelTransaction = bridge.getTopLevelTransactionBoundToThisThread( false );
        return kernelTransaction == null ? null : new TopLevelTransaction( kernelTransaction, null );
    }
}
